src/main/scala/is/hail/expr/package.scala

```diff
@@ -21,6 +22,18 @@ package object expr extends HailRepFunctions {
     def copy(): TypedAggregator[S]
   }
 
+  trait RegionValueAggregator extends Serializable {
+    def typ: Type
+
+    def seqOp(region: MemoryBuffer, off: Long, missing: Boolean): Unit
+
+    def combOp(agg2: RegionValueAggregator): Unit
+
+    def result(region: MemoryBuffer): Long
```

Comment:
Do we always want results appended to a region value with the offset returned? I'm struggling with this a little bit because I feel like we shouldn't be storing /can't currently store pointers to Ints, etc., which it sounds like is what would be happening here? Is there another layer of region value that deals with this?

Reply:
oh wait, never mind. it looks like you're just kind of using these as scratch spaces for an eventual call to addRegionValue, like in the ZippedAggregator?

Reply:
I definitely return a pointer to something like an `Int`. I don't really have any other way of generically referring to a value. Type parameters and any other scala attempts at this will box.

The only current user is ZippedAggregator which copies the primitive (or non-primitive, as the case may be) into the appropriate position in the Struct that collects the many aggregations into one compound datum.

Reply:
@catoverdrive wanna talk in person about this? If your intuition doesn't like this, then maybe I've taken a wrong turn.

Reply:
No, I think this looks fine. The part that I was missing (and correct me if this is wrong) was that you're using these parts of the memory buffer as kind of a scratch space rather than necessarily putting the pointers to results directly in the returned RegionValue.

Reply:
Hmm. I'll be sure to include aggregators in tomorrow's discussion, but let me give you the overview:

To compile an IR expression that uses aggregation, it is first split into a "stage0 IR" and "stage1 array of aggregators". The "stage0 IR" does not reference aggregators nor aggregables. For example:

```
2 * (gs.map(g => g.gt).sum() + gs.count())
```

compiles into a stage0 IR:

```
ApplyBinaryOp(Multiply(),
  I32(2),
  ApplyBinaryOp(Add(),
    GetField(In(0), "0"),
    GetField(In(0), "1")))
```

and a stage1 array of aggregators, which using some pseudo syntax, implement this::

```
[ RegionValueTransformedAggregator(GetField(In(0), "gt"), RegionValueSumAggregator())
, RegionValueCountAggregator()
]
```

So the "RVTransformedAggregator" will create some garbage in the region, however, the base aggregators do not write anything into the region when their `seqOp` and `combOp` methods are called. They store the state (in this case a sum and a count) off-region (in the JVM heap).

After we aggregate over the genotype array (whether it's distributed, as in the column-case, or local in the row-case), we have the desired results, but they are stored in the JVM heap. We need to communicate those values to the stage0 IR. We do this by calling `result` on each aggregator, which will write it's value into the region. Then we create a struct whose fields are "0", "1", ..., and whose values are the results of the 0th, 1st, ... aggregator. (Aside: Unfortunately, this copies the values from their previous location in the region into the struct, because structs are always in-line.) Finally, that struct is passed to the stage0 IR as its second argument (the first argument is the region). The stage0 IR can then reference aggregation results by referencing fields of that struct. For more detail on how that works, check out the aggregators PR #2515 .

Reply:
PS: RVTransformedAggregator:

This one compiles a fragment of IR and uses it as pre-processing step before handing a value to the base aggregator. For example:

```
gs.map(g => g.gt).map(x => 2 * x)
```

Will have a region value transformed aggregator that looks roughly like this:
```
ApplyBinaryOp(Multiply(), I32(2), GetField(In(0), "gt"))
```

Then that gets compiled into JVM byte code and is pre-composed with the seqOp of the base aggregator.

I haven't implemented filtering yet. I guess I should do that with the IR aggregator PR.